orderbook/src/orderbook.rs

```diff
@@ -39,11 +49,42 @@ impl Orderbook {
     }
 
     pub async fn get_orders(&self, filter: &OrderFilter) -> Result<Vec<Order>> {
-        self.storage.get_orders(filter).await
+        let orders_without_balance = self.storage.get_orders(filter).await?;
```

Comment:
Both of the steps in this function could be done more efficiently by moving some functionality into the balance fetcher.
Currently we do:
1. for every order: get_balance, if none put into vec
2. register_many
3. for every order: get_balance

Instead if there was a function like `make_sure_subscription_exits_and_create_and_update_if_not(impl Iteratior<(H160, H160)>` we could avoid locking the mutex for every order in the first step.
Or taking it a step further there could be a function `update_orders_with_balances_and_subscribe(&mut [Order])` that combines everything.
I know that specializes the balance fetcher a lot but I think it's worth it.

This can happen in a future PR and I don't mind implementing it myself. Let me know if I should.

Reply:
I'm personally not really worried that acquiring a mutex will be the bottleneck in this operation. I agree it's nicer to write performant code (e.g. in the case of the balance fetcher itself where the change was a detail of them implementation), but here I feel it's a tradeoff against making a component very specialised for one use case.

Even if we are not going to use this component elsewhere, I personally prefer a simpler interface (here I can register a bunch of subscriptions, potentially unregister them, and query a single balance in a non-blocking form) over a highly specialised one. 
My somewhat subjective argument is that it's cognitively easier to work with/maintain such components in the long run (when we have many different components).

Reply:
My pov is that there is a specific operation that we want to accomplish here: for every order in a vector we want to set the balance which either involves getting it from the cache or from the node.
Based on the specific design of the cache there is a most efficient way to do this which is to calculate the set of uncached balances, fetch them into the cache, update all orders from cache. Doing it this way locks the mutex twice no matter how many orders there are.
But now we are artificially limiting us to a much less efficient implementation because of the way we picked the abstraction boundary interface as it only supports getting the balance of single orders (and register_many). "much less efficient" because for every order we lock the mutex twice and there can be a lot of orders.
To me this looks like a problem with the interface. Why wouldn't it support the exact operation we have created it for? We have generalized prematurely. I agree that it is a more complex interface but the total complexity of the component is still low.
We have this operation that updates the balances in a Vec either way. It can either be here and inefficient or directly in the component and efficient.

Reply:
Maybe my understanding is incorrect, but according to [this benchmark](https://github.com/matklad/lock-bench) acquiring a mutex in rust takes 1.5 nanoseconds in light contention (still less than 10 ns in heavy contention). 

So maybe this question comes down to “premature generalisation” vs “premature optimisation”. Personally, I find generalisation still valuable even if the component is not reused since it allows the maintainer to think in more common/generic concepts (similar to why people tend to use design patterns over and over - others have seen similar components before and thus understand them more quickly).

I originally designed `get_balance` to be non-blocking in order to ensure the read path is fast (as in _"no network request has to be made"_ fast, thus thinking in milliseconds). However, seeing that orders can come from disk and might not yet be registered this might have been unreasonable.
Therefore, I’d be fine to change the interface to a single `async get_balances(impl Iter<(Address, Token)>)` (maybe even with the fact that it creates subscriptions being implicit unless we actively unsubscribe from the orderbook). This way it would still be simple yet more natural with the underlying implementation. Let's do this in a separate PR though.